packages/core/src/tracing/clsHooked/async_local_storage_context.js

```diff
@@ -0,0 +1,324 @@
+   * been set earlier in this context.
+   * @param {string} key
+   * @param {boolean} [fallbackToSharedContext=false]
+   */
+  get(key, fallbackToSharedContext = false) {
+    const activeContext = storage.getStore() || (fallbackToSharedContext ? this.sharedContext : undefined);
+    if (!activeContext) {
+      return undefined;
+    }
+
+    return activeContext[key];
+  }
+
+  /**
+   * Creates a new CLS context in this namespace.
+   * @returns {InstanaCLSContext}
+   */
+  createContext() {
+    const activeContext = storage.getStore();
+    // Prototype inherit existing context if created a new child context within existing context.
+    const context = Object.create(activeContext || Object.prototype);
+    context._ns_name = this.name;
+    context.id = executionAsyncId();
+    return context;
+  }
+
+  /**
+   * Runs a function in a new CLS context. The context is left after the function terminates. Asynchronous work started
+   * in this function will happen in that new context. The return value from that function (if any) is discarded.
+   * If you aren't 100% certain that the function never returns a value or that client code never relies on that value,
+   * use runAndReturn instead.
+   * @param {Function} fn
+   * @param {InstanaCLSContext} ctx
+   */
+  run(fn, ctx) {
+    const context = ctx || this.createContext();
+
+    storage.run(context, () => fn(context));
+    return context;
+  }
+
+  /**
+   * Runs a function in a new CLS context and returns its return value. The context is left after the function
+   * terminates. Asynchronous work started in this function will happen in that new context.
+   * @param {Function} fn
+   * @param {InstanaCLSContext} ctx
+   */
+  runAndReturn(fn, ctx) {
+    let value;
+    this.run((/** @type {InstanaCLSContext} */ context) => {
+      value = fn(context);
+    }, ctx);
+    return value;
+  }
+
+  /**
+   * Runs a function which returns a promise in a new CLS context and returns said promise. The context is left
+   * as soon as the the promise resolves/is rejected. Asynchronous work started in this promise will happen in that new
+   * context.
+   *
+   * If the given function does not create a then-able, an error will be thrown.
+   *
+   * This function assumes that the returned promise is CLS-friendly or wrapped already.
+   * @param {Function} fn
+   * @param {InstanaCLSContext} ctx
+   */
+  runPromise(fn, ctx) {
+    const context = ctx || this.createContext();
+
+    /**
+     * We need the new context to be set as the active context before the promise is called.
+     * Otherwise, the current active context will cause the new span transmission to fail.
+     * Unlike our old code, AsyncLocalStorage.run does not update the active context, so we have to do it manually with
+     * storage.enterWith. This is done by `this.enter`, which calls `enterWith`.
+     * The same happens in the old code, and for the same reasons.
+     */
+    this.enter(context);
```

Comment:
Please, @basti1302 let me know if this change and explanation in the comments satisfies your questioning here: https://github.com/instana/nodejs/pull/416#discussion_r760927718

The call to `enterWith` stays, but the code was cleaned a lot and I no longer update the context twice.

Reply:
Yep, that makes sense and looks good 👍